src/metabase/api/common.clj

```diff
@@ -261,16 +262,49 @@
                                   (ns-name *ns*) fn-name)))
       (assoc parsed :fn-name fn-name, :route route, :docstr docstr))))
 
+(defn validate-param-values
+  "Log a warning if the request body contains any parameters not included in `expected-params` (which is presumably
+  populated by the defendpoint schema)"
+  [{method :request-method route :compojure/route-context body :body} expected-params]
+  (let [extraneous-params (set/difference (set (keys body))
+                                          (set expected-params))]
+    (when (seq? extraneous-params)
+      (log/warn (format (str "Unexpected parameters at %s %s: %s\nPlease add them to the schema or remove them from the"
+                             " API client")
+                        method route (vec extraneous-params))))))
+
+
+(defn method-symbol->keyword
+  "Convert Compojure-style HTTP method symbols (PUT, POST, etc.) to the keywords used internally by
+  Compojure (:put, :post, ...)"
+  [method-symbol]
+  (-> method-symbol
+      name
+      u/lower-case-en
+      keyword))
+
 (defmacro defendpoint*
   "Impl macro for [[defendpoint]]; don't use this directly."
-  [{:keys [method route fn-name docstr args body]}]
+  [{:keys [method route fn-name docstr args body arg->schema]}]
   {:pre [(or (string? route) (vector? route))]}
-  `(def ~(vary-meta fn-name
-                    assoc
-                    :doc          docstr
-                    :is-endpoint? true)
-     (~(symbol "compojure.core" (name method)) ~route ~args
-      ~@body)))
+  (let [method-kw      (method-symbol->keyword method)
+        allowed-params (keys arg->schema)
+        prep-route     #'compojure/prepare-route]
```

Comment:
doesn't feel great grabbing that private fn, but despite being private it seems unlikely to change. And if it ever does the entire app will stop working (and most tests will fail) and we'll know instantly.

Reply:
Or it changes subtly, so as to break production only. 😁